risc0/zkvm/src/groth16.rs

```diff
@@ -0,0 +1,802 @@
+        prepared_inputs
+            .serialize_uncompressed(&mut prepared_inputs_bytes)
+            .map_err(|err| anyhow!(err))?;
+
+        Ok(Self {
+            pvk: pvk_bytes,
+            proof: proof_bytes,
+            prepared_inputs: prepared_inputs_bytes,
+        })
+    }
+
+    /// Computes the prepared verifying key used by Bonsai
+    #[cfg(not(target_os = "zkvm"))]
+    pub fn pvk() -> Result<PreparedVerifyingKey<Bn254>, Error> {
+        let alpha_g1 = g1_from_bytes(&vec![from_u256(ALPHA_X)?, from_u256(ALPHA_Y)?])?;
+        let beta_g2 = g2_from_bytes(&vec![
+            vec![from_u256(BETA_X1)?, from_u256(BETA_X2)?],
+            vec![from_u256(BETA_Y1)?, from_u256(BETA_Y2)?],
+        ])?;
+        let gamma_g2 = g2_from_bytes(&vec![
+            vec![from_u256(GAMMA_X1)?, from_u256(GAMMA_X2)?],
+            vec![from_u256(GAMMA_Y1)?, from_u256(GAMMA_Y2)?],
+        ])?;
+        let delta_g2 = g2_from_bytes(&vec![
+            vec![from_u256(DELTA_X1)?, from_u256(DELTA_X2)?],
+            vec![from_u256(DELTA_Y1)?, from_u256(DELTA_Y2)?],
+        ])?;
+
+        let ic0 = g1_from_bytes(&vec![from_u256(IC0_X)?, from_u256(IC0_Y)?])?;
+        let ic1 = g1_from_bytes(&vec![from_u256(IC1_X)?, from_u256(IC1_Y)?])?;
+        let ic2 = g1_from_bytes(&vec![from_u256(IC2_X)?, from_u256(IC2_Y)?])?;
+        let ic3 = g1_from_bytes(&vec![from_u256(IC3_X)?, from_u256(IC3_Y)?])?;
+        let ic4 = g1_from_bytes(&vec![from_u256(IC4_X)?, from_u256(IC4_Y)?])?;
+        let gamma_abc_g1 = vec![ic0, ic1, ic2, ic3, ic4];
+
+        let vk = VerifyingKey::<Bn254> {
+            alpha_g1,
+            beta_g2,
+            gamma_g2,
+            delta_g2,
+            gamma_abc_g1,
+        };
+
+        Ok(prepare_verifying_key(&vk))
+    }
+
+    /// Verifies the Groth16 instance
+    pub fn verify(&self) -> Result<(), Error> {
+        let pvk = &PreparedVerifyingKey::deserialize_uncompressed(&*self.pvk)
+            .map_err(|err| anyhow!(err))?;
+        let proof = &Proof::deserialize_uncompressed(&*self.proof).map_err(|err| anyhow!(err))?;
+        let prepared_inputs = &G1Projective::deserialize_uncompressed(&*self.prepared_inputs)
+            .map_err(|err| anyhow!(err))?;
+        match ark_Groth16::<Bn254>::verify_proof_with_prepared_inputs(pvk, proof, prepared_inputs)
+            .map_err(|err| anyhow!(err))?
+        {
+            true => Ok(()),
+            false => bail!("Invalid proof"),
+        }
+    }
+
+    /// Serialize the Groth16 instance into Vec<u8>
+    pub fn to_vec(&self) -> Vec<u8> {
+        let mut result = Vec::new();
+        result.extend(&self.pvk);
+        result.extend(&self.proof);
+        result.extend(&self.prepared_inputs);
+        result
+    }
+
+    /// Compute the SHA256 digest of the Groth16 instance
+    #[cfg(not(target_os = "zkvm"))]
+    pub fn digest(&self) -> Digest {
+        let hash = Impl::hash_bytes(&self.to_vec());
+        *hash.as_ref()
+    }
```

Comment:
Specifically, this is the hash of a particular encoding. I would suggest removing this method and making the encoding explicit where you need a hash.

```suggestion
```